File: Ruby/Deepspace/lib/Weapon.rb
#encoding:utf-8
module Deepspace
class Weapon
  attr_reader :name, :type, :uses
  def initialize(name, type, uses)
    @name = name
    @type = type
    @uses = uses     
  end
  
  def self.newCopy(copy)
    Weapon.new(copy.name, copy.type, copy.uses) 
  end
  
  def power
    return type.power
  end
  
  def useIt
    if @uses > 0
      @uses -= 1
      return power
    else
      return 1.0
    end
  end
  
  def to_s
   "\nname #{@name}, type #{@type} , uses #{@uses}"

  end
  
  def getUIversion
    WeaponToUI.new(self)
  end
end
end

File: Java/DeepSpace/src/deepspace/Weapon.java
package deepspace;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author victor
 */
public class Weapon {
    private String name;
    private WeaponType type;
    private int uses;

    Weapon(String name, WeaponType type, int uses) {
        this.name = name;
        this.type = type;
        this.uses = uses;
    }
    
    Weapon(Weapon w) {
        name = w.name;
        type = w.type;
        uses = w.uses;
    }
    
    WeaponToUI getUIversion() {
        return new WeaponToUI(this);
    }

    public WeaponType getType() {
        return type;
    }

    public int getUses() {
        return uses;
    }
    
    public float power(){
        return type.getPower();
    }
    
    public float useIt(){
        if(uses > 0){
            uses--;
            return power();
        } else
            return 1.0f;
    }

    @Override
    public String toString() {
        return "Weapon{" + "name=" + name + ", type=" + type + ", uses=" + uses + '}';
    }
    
    
}
